### PR TITLE
Create a BDD Test Automation Framework with ginkgo

### PR DESCRIPTION
**Pull request by alejovicu**
* Create a BDD test automation framework using ginkgo and gomega to test the asset generation before launch the cluster installation (https://jira.coreos.com/browse/CORS-923)
* Add a test suite for `openshift-install create install-config`for libvirt platform (https://jira.coreos.com/browse/CORS-858)

### CODE REVIEW

**Comment by wking** [comment]
Do we need this?  For the libvirt tests, I was hoping to run the usual origin e2e suite.  And if you're only concerned about exercising `create install-config`, a simpler approach would be to add a `cmd/openshift-install/create_test.go` for our existing Go unit job.

**Comment by alejovicu** [comment]
Hit @wking  
This framework should cover the entire application. 
At this moment is just the setup of the framework and the test suite for "create install-config" command. but It will increase adding the test suites for "_create ignition-configs,  create manifests, create cluster, destroy-cluster_...". When the cluster is deployed, the usual origin e2e suite will be executed to certify the cluster health.

Now, in the test workflow, the idea is to run the config tests before the cluster health tests because some issues can be found without the cluster deploy saving some time.

File: tests/bdd-smoke/Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[[constraint]]
```
Comment (crawford) [inline tests/bdd-smoke/Gopkg.toml]: Can you break out the dependency imports into another commit? It just makes it a little easier to follow the history.

File: tests/bdd-smoke/suites/config/libvirt/install_config_test.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+			Ω(openshiftInstallerPath + "/.openshift_install_state.json").Should(BeAnExistingFile())
+		})
+		Specify("And I expect install-config.yml file contains the info specified in the environment variables", func() {
+			utils.ValidateInstallConfig("../../../../../install-config.yml")
+		})
+		Specify("And I expect .openshift_install_state.json file contains the info specified in the environment variables", func() {
+			utils.ValidateInstallStateConfig("../../../../../.openshift_install_state.json")
+		})
+		Specify("And I expect to get the debug output log from the console", func() {
+			Ω(cmdOut).Should(ContainSubstring(`level=debug msg="Fetching \"Install Config\"..."`))
```
Comment (crawford) [inline tests/bdd-smoke/suites/config/libvirt/install_config_test.go]: This might be too verbose to check. It's probably okay for now.

File: tests/bdd-smoke/utils/assertion.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package utils
+
+import (
+	"log"
+	"strings"
+)
+
+// AssertStringToContain compare two strings and fails if aren't equal
```
Comment (crawford) [inline tests/bdd-smoke/utils/assertion.go]: This comment isn't accurate.

File: tests/bdd-smoke/utils/assertion.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+func AssertStringToContain(mainString string, subString string) {
```
Comment (crawford) [inline tests/bdd-smoke/utils/assertion.go]: I would name this `AssertStringContains`.

File: tests/bdd-smoke/Gopkg.toml (hunk 1 above)
Comment (alejovicu) [reply tests/bdd-smoke/Gopkg.toml]: Done

**Comment by alejovicu** [comment]
Hi @abhinavdahiya,
The split in 3 commits is done.
I created this as a separate vendor because this framework uses dependencies that only apply for it, like gomega that would be useless by other modules in the installer.

File: hack/run-bdd-suite.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#! /usr/bin/sh
+# Example: ./hack/run-bdd-suite.sh tests/bdd-smoke/suites/config/libvirt
+
+if [ "$IS_CONTAINER" != "" ]; then
+  mkdir ~/.ssh
+  echo "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQCc4amSo5Z59mH4bUgZ4M8A1sURg+qUKkYjZ9+0fft5/OmSbe34Ge9gUrJ8UKNpmbC+W26v9qKX21NCYct1QOgIlVqyIjZUll0NfcJXToAd1o2GrItMlpQEpXfQGNOjZtWu1uyO187qbGMwnq8e8EH9R1IJZwTI6sOVnc8sSyzydQ==" > ~/.ssh/id_rsa.pub
```
Comment (abhinavdahiya) [inline hack/run-bdd-suite.sh]: hmm, what is this for ?

File: hack/run-bdd-suite.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  
+  TAGS=libvirt ./hack/build.sh
+  go get github.com/onsi/ginkgo/ginkgo
+  ginkgo "${@}"
```
Comment (abhinavdahiya) [inline hack/run-bdd-suite.sh]: is it possible to use ginko from the vendor?

File: hack/run-bdd-suite.sh (hunk 5 above)
Comment (alejovicu) [reply hack/run-bdd-suite.sh]: this is to add a valid ssh test key inside of the golang's container and then validate it by the framework.

File: hack/run-bdd-suite.sh (hunk 6 above)
Comment (alejovicu) [reply hack/run-bdd-suite.sh]: I think so, the ginkgo binary should be compiled inside of the container, I will check how to do this.

Comment (alejovicu) [reply hack/run-bdd-suite.sh]: @abhinavdahiya done

File: tests/bdd-smoke/utils/command_runner.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+func RunOSCommandWithArgs(command string, arguments []string, path string) string {
+	cmd := exec.Command(command, arguments...)
+	cmd.Dir = path
+	cmd.Stdin = strings.NewReader("")
+	var outb, errb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	err := cmd.Run()
+	if err != nil {
+		fmt.Println("Error: failed running the command \"" + command + "\" with the arguments:")
```
Comment (wking) [inline tests/bdd-smoke/utils/command_runner.go]: nit: this could be more robust and compact with:

```go
fmt.Printf("Error: failed running the command %q with the arguments:", command)
```

although printing straight to the terminal feels strange anyway.  Does that get captured somewhere?  Should we be logging these instead, or attaching it to the error before passing it to `log.Fatal`?

File: tests/bdd-smoke/utils/libvirt_inst_state_file_validator.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+		log.Fatalf("Failed while open file: %v", err)
+	}
+
+	defer openshiftInstState.Close()
+	byteValue, _ := ioutil.ReadAll(openshiftInstState)
+
+	var installState installStateConf
+	json.Unmarshal(byteValue, &installState)
+
+	AssertStringContains(installState.InstallConfig.Config.BaseDomain,
```
Comment (wking) [inline tests/bdd-smoke/utils/libvirt_inst_state_file_validator.go]: This and later parts of `ValidateInstallStateConfig` feel like a hand-crafted special case of [`reflect.DeepEqual`][1].  Is it doing something else?

[1]: https://golang.org/pkg/reflect/#DeepEqual

File: tests/bdd-smoke/utils/libvirt_inst_state_file_validator.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+		"*installconfig.baseDomain.BaseDomain not found in "+fileName)
+
+	AssertStringContains(installState.InstallConfig.Config.Metadata.Name,
+		expectedData.ClusterName,
+		"*installconfig.config.metadata.name not found in "+fileName)
+	AssertStringContains(installState.ICClusterName.ClusterName,
+		expectedData.ClusterName,
+		"*installconfig.clusterName.ClusterName not found in "+fileName)
+
+	// OPENSHIFT_INSTALL_PLATFORM is implicitly validated by next checks. If its not Libvirt platform, the Name and URI will not be accessible
```
Comment (wking) [inline tests/bdd-smoke/utils/libvirt_inst_state_file_validator.go]: nit: This references the environment variable dropped in 6be4c253e (#861).
